### ISSUE TITLE
Add ability to condition a request with one or more lock resourceIds

### ISSUE THREAD

**Issue by alechenninger** [lightblue-platform/lightblue-core#569]
That is, I make a request (like a data update or whatever), and with the body of the request I provide information about locks required to make the request. The, lightblue should only perform the request if provided locks are acquired at the time the request is about to be processed, and guarantee to keep the locks acquired at least until the request is finished. If the locks are not all acquired when the request is received, fail the request with a specific error about the required locks not being acquired.

This will allow for some atomic operations.

Today, even if you ping a lock just before some request, together they are not truly atomic, because technically a lock with TTL can still be released before the request is processed. Obviously if the TTL is long enough you can make this very unlikely, but I'd be more comfortable with "not even possible." It's just less cognitive overhead when you know you don't have to worry about something :). Plus, there are downsides to long TTLs, too.

Non-TTL locks are not really practical since there's no telling what could happen to the code, jvm, vm, whatever, that's supposed to eventually release it.

**Comment by bserdar** [comment]
Unsafe. TTL is required. Locking isn't a way to emulate transactions. Best option is to wait until some sort of transaction support is added. The first stage is probably going to be addition of rollback support. It may not be possible to add atomic transaction support.

**Comment by alechenninger** [comment]
Can you elaborate on unsafe? I agree TTL would have to be required; its only really useful when using TTLs.

**Comment by bserdar** [comment]
Packaging a lock and update operation into a single request won't change how it is run. It can still lock, start a series of updates, and fail in the middle, leaving data in inconsistent state. There is no benefit in packaging the lock request and updates into one request.

**Comment by alechenninger** [comment]
How do you mean, "fail in the middle?" Do you mean lose the lock (because TTL times out)? That would be a requirement of the feature: to make the lock live until the request is done. If there is some other failure, like a bad request, for my use case nothing else would change. It should still ensure nothing else can get that lock until the request is done (at least, of course the TTL could be longer than the request processing).

**Comment by bserdar** [comment]
How do you make the lock live until the request is done, if the request fails for, say, network error? To put it another way: how do you tag the lock to notify that it's ttl has started if there is a network failure?

**Comment by alechenninger** [comment]
Hrm. Well, I guess it depends on where the network error is. If lightblue never received the request, my intuition would be that because there was no request, there would be no guarantee on maintaining the lock -- there is no request with which to guarantee. It would be like you didn't make the request at all. If there was any failure after starting the request (and the client had the required lock) such as failure to talk to mongo, then the request should only guarantee maintaining the lock up until it failed. Again, it wouldn't mean that the lock would be necessarily lost (or not) after the request, it is only guaranteed to still be acquired up until it is done, whether it finished with a failure or not.

I realize I'm pretty ignorant to how locks are implemented currently, and I am guessing implementing this would probably require some work, but I find this use case very useful assuming it is not flawed in some way that I still don't see.

I guess this is what I would like to be able to do:

1. Acquire a lock with 30 second TTL, say
2. If I acquire it, make a request, requiring above lock, and ensure it is maintained until the request is done (whether it finishes successfully or not). This may mean that the lock, in rare cases, lives past its TTL.
3. The lock is lost once TTL is up, or is lost when the request is done if the TTL was up before the request was done.

This would give good confidence that you would have no race condition where you acquire a lock with TTL, but by the time your request is sent and processed somehow your TTL is up. The longer your TTL, the less likely this is to happen, but it'd be nice to have that guarantee.

Am I missing something?

**Comment by bserdar** [comment]
The problem is in how TTL is implemented. Locks have expiration dates, set
using now+TTL. Once current time passes TTL, lock is assumed to be
released. So, if you want to keep lock alive during request processing, you
essentially need to have infinite TTL, and then set the TTL once processing
is complete. If processing fails before the TTL is set, you end up with a
lock that never expires.

**Comment by alechenninger** [comment]
Yeah hrm, this is basically the same problem you have from client side as well. Interesting. I was hoping there was a way lightblue could get around this internally. Thanks!

**Comment by bserdar** [comment]
There isn't. The only synchronization feature we have with mongo is atomic
single doc updates, and reliability of unique indexes. So locking is based
on updates and unique indexes.

**Comment by alechenninger** [comment]
As a much simpler compromise, could we have something like this?

1. I acquire a lock with TTL
2. I make a request with resourceId and my clientId, request returns error if at time client does not have lock. Otherwise, pings the lock and does the request.
3. Everything else happens as normal (no special lock guarantees or anything). Which means TTL could theoretically timeout before request is done but, it's so unlikely if you have a reasonably long TTL it's probably not a serious issue.

I can do this today client side, so similar to my other request, it would be nice if I could push some of this server side, which also reduces some of the latency between the ping and the subsequent request.

**Comment by bserdar** [comment]
Possible. What you really need is continuous pings of the lock as long as
operation is in progress. Can you put this into a separate ticket? I'll
think about this to make sure we're not missing anything.

**Comment by alechenninger** [comment]
> What you really need is continuous pings of the lock as long as operation is in progress

This is exactly what I'm toying around with implementing client side, but it's not totally essential IMHO

**Comment by bserdar** [comment]
Wait, I don't think this is still safe. There is no guarantee that the lock
will be kept until a db call is complete. So, this is simply a convenience
that can be implemented at client side.

**Comment by alechenninger** [comment]
Yes, exactly. Convenience, and a little bit better performance (one call for ping+request instead of two calls).
